### PR TITLE
feat: Use of configurable concurrency for extracting assets from BigQuery

### PR DESCRIPTION
**Pull request by Mayurjag**

### CODE REVIEW

File: plugins/extractors/bigquery/bigquery.go [hunk 1]
```diff
@@ -302,37 +325,51 @@ func (e *Extractor) extractTable(ctx context.Context, ds *bigquery.Dataset, emit
...
+						attribute.String("bq.project_id", e.config.ProjectID),
+						attribute.String("bq.dataset_id", ds.DatasetID),
+					))
+					e.logger.Debug("excluding table from bigquery extract", "dataset_id", ds.DatasetID, "table_id", table.TableID)
+					return nil
+				}
 
-			tableFQN := table.FullyQualifiedName()
+				mu.Lock()
+				defer mu.Unlock()
```
Comment (haveiss) [inline plugins/extractors/bigquery/bigquery.go]: why we need mutex here?
let's use channel to managing the concurrency

File: plugins/extractors/bigquery/bigquery.go [hunk 2]
```diff
@@ -204,6 +206,8 @@ func (e *Extractor) Init(ctx context.Context, config plugins.Config) error {
 		e.logger.Error("failed to create policy tag manager client", "err", err)
 	}
 
+	e.concurrency = e.config.Concurrency
```
Comment (batrov) [inline plugins/extractors/bigquery/bigquery.go]: where is this being used?

File: plugins/extractors/bigquery/bigquery.go [hunk 3]
```diff
@@ -219,17 +223,27 @@ func (e *Extractor) Extract(ctx context.Context, emit plugins.Emit) error {
 			return err
 		}
 
+		ch := make(chan *bigquery.Dataset, len(datasets))
+
+		numWorkers := 5
```
Comment (batrov) [inline plugins/extractors/bigquery/bigquery.go]: should be configurable

File: plugins/extractors/bigquery/bigquery.go [hunk 4]
```diff
@@ -301,38 +332,50 @@ func (e *Extractor) extractTable(ctx context.Context, ds *bigquery.Dataset, emit
...
-			}
-
-			asset, err := e.buildAsset(ctx, table, tmd)
-			if err != nil {
-				e.logger.Error("failed to build asset", "err", err, "table", tableFQN)
-				continue
-			}
-
-			emit(models.NewRecord(asset))
+			wg.Add(1)
```
Comment (batrov) [inline plugins/extractors/bigquery/bigquery.go]: we can init the wg size before the loop, using tables length

File: plugins/extractors/bigquery/bigquery.go (hunk 3 above)
Comment (haveiss) [inline plugins/extractors/bigquery/bigquery.go]: Could you please explain why we hardcoded it to 5? I don't see the concurrency variable being used anywhere

File: plugins/extractors/bigquery/bigquery.go [hunk 5]
```diff
@@ -301,38 +329,50 @@ func (e *Extractor) extractTable(ctx context.Context, ds *bigquery.Dataset, emit
...
 		if !hasNext {
 			break
 		}
 	}
+
+	return nil
+}
+
+func (e *Extractor) processTable(ctx context.Context, ds *bigquery.Dataset, table *bigquery.Table, emit plugins.Emit, wg *sync.WaitGroup) {
+	defer wg.Done()
```
Comment (batrov) [inline plugins/extractors/bigquery/bigquery.go]: should be consistent with the processDatasetWorker approach, can move this outside the processTable function

File: plugins/extractors/bigquery/bigquery.go [hunk 6]
```diff
@@ -219,17 +220,27 @@ func (e *Extractor) Extract(ctx context.Context, emit plugins.Emit) error {
 			return err
 		}
 
+		ch := make(chan *bigquery.Dataset, len(datasets))
```
Comment (haveiss) [inline plugins/extractors/bigquery/bigquery.go]: just parallel all the dataset, and use the buffered channel to process the concurrency of the tables

**Review by haveiss — APPROVED** [review]

**Review by haveiss — APPROVED** [review]

**Pull request merged**